core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java

```diff
@@ -48,54 +72,165 @@
+                }
+            }
+        }
+        this.inclusive = inclusive;
+        this.sortedRanges = sortedRanges;
     }
 
     static SortedRangeSet none(Type type)
     {
-        return copyOf(type, Collections.emptyList());
+        return new SortedRangeSet(
+                type,
+                new boolean[0],
+                // TODO This can perhaps use an empty block singleton
+                type.createBlockBuilder(null, 0).build());
     }
 
     static SortedRangeSet all(Type type)
     {
-        return copyOf(type, Collections.singletonList(Range.all(type)));
+        return new SortedRangeSet(
+                type,
+                new boolean[] {false, false},
+                // TODO This can perhaps use a "block with two nulls" singleton
+                type.createBlockBuilder(null, 2)
+                        .appendNull()
+                        .appendNull()
+                        .build());
+    }
+
+    @Deprecated // For JSON deserialization only
+    @JsonCreator
+    public static SortedRangeSet fromJson(
+            @JsonProperty("type") Type type,
+            @JsonProperty("inclusive") boolean[] inclusive,
+            @JsonProperty("sortedRanges") Block sortedRanges)
+    {
+        if (sortedRanges instanceof BlockBuilder) {
+            throw new IllegalArgumentException("sortedRanges must be a block: " + sortedRanges);
+        }
+        return new SortedRangeSet(type, inclusive.clone(), sortedRanges);
     }
 
     /**
      * Provided discrete values that are unioned together to form the SortedRangeSet
      */
     static SortedRangeSet of(Type type, Object first, Object... rest)
     {
-        List<Range> ranges = new ArrayList<>(rest.length + 1);
-        ranges.add(Range.equal(type, first));
+        if (rest.length == 0) {
+            return of(type, first);
+        }
+
+        MethodHandle comparisonOperator = TUPLE_DOMAIN_TYPE_OPERATORS.getComparisonOperator(type, simpleConvention(FAIL_ON_NULL, BLOCK_POSITION, BLOCK_POSITION));
+
+        BlockBuilder blockBuilder = type.createBlockBuilder(null, 1 + rest.length);
+        checkNotNaN(type, first);
+        writeNativeValue(type, blockBuilder, first);
         for (Object value : rest) {
-            ranges.add(Range.equal(type, value));
+            checkNotNaN(type, value);
+            writeNativeValue(type, blockBuilder, value);
         }
-        return copyOf(type, ranges);
+        Block block = blockBuilder.build();
+
+        List<Integer> indexes = new ArrayList<>(block.getPositionCount());
+        for (int position = 0; position < block.getPositionCount(); position++) {
+            indexes.add(position);
+        }
+        indexes.sort((left, right) -> compareValues(comparisonOperator, block, left, block, right));
+
+        int[] dictionary = new int[block.getPositionCount() * 2];
+        dictionary[0] = indexes.get(0);
+        dictionary[1] = indexes.get(0);
+        int dictionaryIndex = 2;
```

Comment:
> Oh the rewritten block would by 2*long. That is the motivation, right?

yes

> instead just rewriting the block with values using computed order?

note that the values are _not_ in order in the block

i put them in _input_ order into a block, sort the indexes, and create sorted view on top of the unsorted data

alternatively i could rewrite the block again (in order; doubling, or still keeping dictionary on top)

i am not convinced this is optimal; i think it may not matter actually